libraries/AP_Generator/AP_Generator_IE_FuelCell.h

```diff
@@ -0,0 +1,178 @@
+        BAT_OT                  = 0x10000000,//(1 << 28),
+        NO_FAN                  = 0x8000000,//(1 << 27),
+        FAN_OVERRUN             = 0x4000000,//(1 << 26),
+        STACK_OT1_2             = 0x2000000,//(1 << 25),
+        STACK_OT2_2             = 0x1000000,//(1 << 24),
+        BAT_UV_2                = 0x800000,//(1 << 23),
+        BAT_OT_2                = 0x400000,//(1 << 22),
+        START_TIMEOUT           = 0x200000,//(1 << 21),
+        STOP_TIMEOUT            = 0x100000,//(1 << 20),
+        START_UNDER_PRESSURE    = 0x80000,//(1 << 19),
+        TANK_UNDER_PRESSURE     = 0x40000,//(1 << 18),
+        TANK_LOW_PRESSURE       = 0x20000,//(1 << 17),
+    };
+
+    struct Lookup_FS_IE6 {
+        ERR_CODE_IE6 option;
+        const uint32_t bit_mask;
+    };
+
+    static const Lookup_FS_IE6 lookup_fs_ie6[];
+
+    // IE 2.4kW failsafes and corresponding bitmask lookup table
+    enum class ERR_CODE_IE24 {
+        MINOR_INTERNAL = 1, // Minor internal error is to be ignored
+        REDUCED_POWER = 10,
+        SPM_LOST = 11,
+        PRESSURE_LOW = 20,
+        BATTERY_LOW = 21,
+        PRESSURE_ALERT = 30,
+        START_DENIED = 31,
+        SYSTEM_CRITICAL = 32,
+        PRESSURE_CRITICAL = 33,
+        BATTERY_CRITICAL = 40,
+    };
+
+    struct Lookup_FS_IE24 {
+        ERR_CODE_IE24 option;
+        const uint32_t bit_mask;
+    };
+
+    static const Lookup_FS_IE24 lookup_fs_ie24[];
+
+    // IE fuel cell running states and corresponding lookup table
+    enum class FUEL_CELL_STATE {
+        STARTING       = 0,
+        READY          = 1,
+        RUNNING        = 2,
+        FAULT          = 3,
+        BATTERY_ONLY   = 8,
+    };
+
+    struct Lookup_State {
+        FUEL_CELL_STATE option;
+        const char *msg_txt;
+    };
+
+    static const Lookup_State lookup_state[];
+
+    // The fuel cell state structure
+    struct FuelCellState {
+        float                   consumed_mah;          // Consumed current (mAh)
+        uint32_t                failsafe;              // The failsafe state
+        uint32_t                err_code;              // The error code from fuel cell
+        uint32_t                last_err_code;         // The previous error code from fuel cell
+        enum FUEL_CELL_STATE    state;                 // The PSU state
+        enum FUEL_CELL_STATE    last_state;            // The previous PSU state
+        uint32_t                last_time_ms;          // Time we got a reading
+        bool                    healthy;               // Is the driver working
+    } _state;
+
+    // Data that is received over telem and is only being logged
+    struct Log_Only_Data {
+        int16_t                 pwr_out;               // Output power (Watts)
+        uint16_t                spm_pwr;               // SPM power draw (Watts)
+        uint32_t                last_time_ms;          // Time we last wrote to log
+    } _logging;
+
+    // Temporary state params
```

Comment:
Would be nice to put these into a structure, something like `struct parsed_value` or similar.